### PR TITLE
Feature/fix communication request

### PR DESCRIPTION
**Pull request by Filienko**
Make current published CRs not be sent when the user decided to stop receiving Twilio SMS. Cancel the ones that were scheduled for time when the user would be not active as per https://www.pivotaltracker.com/n/projects/2584667

### CODE REVIEW

File: isacc_messaging/api/isacc_record_creator.py [hunk 1]
```diff
@@ -447,11 +481,9 @@ def execute_requests(self) -> Tuple[List[dict], List[dict]]:
 
         return successes, errors
 
-    def process_cr(self, errors, cr, successes):
+    def process_cr(self, cr, successes):
         try:
             status = self.convert_communicationrequest_to_communication(cr=cr)
             successes.append({'id': cr.id, 'status': status})
         except Exception as e:
```
Comment (pbugni) [inline isacc_messaging/api/isacc_record_creator.py]: no point in catching an exception if all we do is re-raise it.  kill the try: except: and let it percolate up.

File: isacc_messaging/api/isacc_record_creator.py [hunk 2]
```diff
@@ -422,13 +434,35 @@ def execute_requests(self) -> Tuple[List[dict], List[dict]]:
 
         for cr_json in next_in_bundle(result):
             cr = CommunicationRequest(cr_json)
+            patient = resolve_reference(cr.recipient[0].reference)
+            patient_unsubscribed = any(
```
Comment (pbugni) [inline isacc_messaging/api/isacc_record_creator.py]: i don't believe you want `any` here.  i think you need to track down the `sms` telecom, and then see if the LAST period in that telecom's array has an end that is < now.

File: isacc_messaging/api/isacc_record_creator.py [hunk 3]
```diff
@@ -376,6 +382,11 @@ def on_twilio_message_received(self, values):
         pt = Patient(pt)
 
         message = values.get("Body")
+        # if the user requested to resubscribe, mark him as active
+        if "start" == message.lower():
+            sms_telecom_entry = next((entry for entry in pt.telecom if entry.system.lower() == 'sms'))
+            sms_telecom_entry.period.end = None
```
Comment (pbugni) [inline isacc_messaging/api/isacc_record_creator.py:388]: I thought we'd decided to add a new period in this case, with `start` set to now, and no `end` defined?  Isn't sms_telecom_entry an array of periods?

File: isacc_messaging/api/isacc_record_creator.py [hunk 4]
```diff
@@ -117,6 +117,12 @@ def convert_communicationrequest_to_communication(self, cr):
             result = self.send_twilio_sms(message=expanded_payload, to_phone=target_phone)
 
         except TwilioRestException as ex:
+            # The error is raised when the user has unscubscribed 
+            # mark this user as inactive for the future CRs
+            sms_telecom_entry = next((entry for entry in patient.telecom if entry.system.lower() == 'sms'))
+            sms_telecom_entry.period.end = FHIRDate(datetime.utcnow().strftime('%Y-%m-%dT%H:%M:%SZ'))
```
Comment (pbugni) [inline isacc_messaging/api/isacc_record_creator.py]: isn't sms_telecom_entry.period an array of Periods?

**Review by pbugni — CHANGES_REQUESTED** [review]
Nice job!  A few concerns to address.

File: isacc_messaging/api/isacc_record_creator.py (hunk 4 above)
Comment (Filienko) [reply isacc_messaging/api/isacc_record_creator.py]: I think it has max cardinality of 1 (so, 1 period?).
image

File: isacc_messaging/api/isacc_record_creator.py (hunk 3 above)
Comment (pbugni) [reply isacc_messaging/api/isacc_record_creator.py:388]: i was wrong, period is a 0..1 cardinality

File: isacc_messaging/api/isacc_record_creator.py (hunk 4 above)
Comment (pbugni) [reply isacc_messaging/api/isacc_record_creator.py]: i was wrong, period is a 0..1 cardinality

File: isacc_messaging/api/isacc_record_creator.py (hunk 3 above)
Comment (Filienko) [reply isacc_messaging/api/isacc_record_creator.py:388]: Resolved

File: isacc_messaging/api/isacc_record_creator.py [hunk 5]
```diff
@@ -117,6 +117,12 @@ def convert_communicationrequest_to_communication(self, cr):
             result = self.send_twilio_sms(message=expanded_payload, to_phone=target_phone)
 
         except TwilioRestException as ex:
+            # The error is raised when the user has unscubscribed
```
Comment (pbugni) [inline isacc_messaging/api/isacc_record_creator.py]: hmm, are you SURE that's what the error was?  i expect that's what is being developed now, but code changes in time.
i think we should check the value of `ex` to verify this assumption.

File: isacc_messaging/api/isacc_record_creator.py [hunk 6]
```diff
@@ -376,6 +382,11 @@ def on_twilio_message_received(self, values):
         pt = Patient(pt)
 
         message = values.get("Body")
+        # if the user requested to resubscribe, mark him as active
```
Comment (pbugni) [inline isacc_messaging/api/isacc_record_creator.py]: remove gender from comment here.  "mark patient as active"

File: isacc_messaging/api/isacc_record_creator.py [hunk 7]
```diff
@@ -422,13 +434,49 @@ def execute_requests(self) -> Tuple[List[dict], List[dict]]:
...
             if cr.occurrenceDateTime.date < cutoff:
-                # anything older than cutoff will never be sent (#1861758)
+                # Anything older than cutoff will never be sent (#1861758)
                 # and needs a status adjustment lest it throws off other queries
                 # like next outgoing message time
                 skipped_crs.append(cr)
                 continue
-            self.process_cr(errors, cr, successes)
+            if patient_unsubscribed:
+                if occurrence_utc < datetime.now(timezone.utc):
```
Comment (pbugni) [inline isacc_messaging/api/isacc_record_creator.py]: i see elsewhere, we use `datetime.now().astimezone()`
probably the same result, but i'm a fan of consistency.

**Review by pbugni — CHANGES_REQUESTED** [review]
solid work.. couple nits

File: isacc_messaging/api/isacc_record_creator.py [hunk 8]
```diff
@@ -117,6 +117,12 @@ def convert_communicationrequest_to_communication(self, cr):
             result = self.send_twilio_sms(message=expanded_payload, to_phone=target_phone)
 
         except TwilioRestException as ex:
+            # In case of unsubcribed patient, mark as unsubscribed
```
Comment (pbugni) [inline isacc_messaging/api/isacc_record_creator.py:120]: nice!  thank you for the comment, as i'd never know what 21610 meant!

**Review by pbugni — APPROVED** [review]
💯 -- thank you.

**Pull request merged**